### PR TITLE
Update Emoji Datasource and fix uniform rendering across platforms

### PR DESCRIPTION
**Pull request by akshayasalvi**
### Details
- Some of the emojis were not showing up properly on all platforms because they were text type emojis. Hence we replaced the datasource.
- This includes adding more emojis to our database too.

### Fixed Issues

$ https://github.com/Expensify/App/issues/10356   
PROPOSAL: https://github.com/Expensify/App/issues/10356#issuecomment-1250226783

### Tests

1. Open any chat
2. Click on the Emoji action on the composer
3. Select a problematic emoji like Coffin or Peace
4. Ensure the emoji is colored in all the platforms (Web, Desktop, mWeb, Mobile apps)
5. Click on the different emojis and ensure they render well
6. Keep submitting 1-2 emojis, and see they're updated in the Frequently used Emojis

- [ ] Verify that no errors appear in the JS console

### QA Steps

1. Open any chat
2. Click on the Emoji action on the composer
3. Select a problematic emoji like Coffin or Peace
4. Ensure the emoji is colored in all the platforms (Web, Desktop, mWeb, Mobile apps)
5. Click on the different emojis and ensure they render well
6. Keep submitting 1-2 emojis, and see they're updated in the Frequently used Emojis

- [ ] Verify that no errors appear in the JS console

#### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

#### Web

Screenshot 2022-10-31 at 11 17 38 AM

#### Mobile Web - Chrome

image

#### Mobile Web - Safari

image

#### Desktop

image

#### iOS

image

#### Android

image

### CODE REVIEW

**Comment by akshayasalvi** [comment]
@tgolen @mananjadhav I have made additional changes so that the `:emojicode:` searching logic works as expected.

Also there is one test failing because we're checking if the emoji matches a regex. I am not sure why we do that? We are maintaining two lists of Emojis - emoji.js and a regex. Do you want me to fix this?

**Comment by akshayasalvi** [comment]
@tgolen @mananjadhav The new emojis might not work with the existing regex, and we will have to update the `isSingleEmoji` function.

Take the following example of `"face_in_clouds:` emoji. The emoji image is 😶‍🌫️ but the regex match is:  `[ '😶', '‍', '🌫']`  a composite emoji.

**Comment by mananjadhav** [comment]
@tgolen @mananjadhav  Apologies for too many messages. I finally found the problem. 

The tests were failing due to the following reasons:
- `isSingleEmoji` wasn't comprehensive enough to cover the composite emojis. Also, this isn't used anywhere in code except for the Tests. So I've removed it.
- Hence, I've replaced all the `isSingleEmoji` calls with `containsOnlyEmojis`.
- There was one fix required for male/female emojis to work with `containsOnlyEmojis`
- The emojis in `EmojiTest` weren't correct, (for `*`, `1`)

**Comment by akshayasalvi** [comment]
@tgolen @mananjadhav  Apologies for too many messages. I finally found the problem. 

The tests were failing due to the following reasons:
- `isSingleEmoji` wasn't comprehensive enough to cover the composite emojis. Also, this isn't used anywhere in code except for the Tests. So I've removed it.
- Hence, I've replaced all the `isSingleEmoji` calls with `containsOnlyEmojis`.
- There was one fix required for male/female emojis to work with `containsOnlyEmojis`
- The emojis in `EmojiTest` weren't correct, (for `*`, `1`)

**Comment by mananjadhav** [comment]
@tgolen @mananjadhav Apologies for too many messages. I finally found the problem.

 The tests were failing due to the following reasons:
 
 * `isSingleEmoji` wasn't comprehensive enough to cover the composite emojis. Also, this isn't used anywhere in code except for the Tests. So I've removed it.
 * Hence, I've replaced all the `isSingleEmoji` calls with `containsOnlyEmojis`.
 * There was one fix required for male/female emojis to work with `containsOnlyEmojis`
 * The emojis in `EmojiTest` weren't correct, (for `*`, `1`)

File: tests/unit/EmojiTest.js [hunk 1]
```diff
@@ -26,38 +27,38 @@ describe('EmojiTest', () => {
...
-        expect(EmojiUtils.isSingleEmoji('👨‍👩‍👦️')).toBe(true);
-        expect(EmojiUtils.isSingleEmoji('👩‍👩‍👧‍👦️')).toBe(true);
+        expect(EmojiUtils.containsOnlyEmojis('👨‍👩‍👦️')).toBe(true);
+        expect(EmojiUtils.containsOnlyEmojis('👩‍👩‍👧‍👦️')).toBe(true);
 
-        // GIVEN an emoji that has a length of 2 (flags) WHEN we check if it's a single emoji THEN it should return true
-        expect(EmojiUtils.isSingleEmoji('🇺🇲')).toBe(true);
-        expect(EmojiUtils.isSingleEmoji('🇮🇳')).toBe(true);
-        expect(EmojiUtils.isSingleEmoji('🇺🇦️')).toBe(true);
+        // // GIVEN an emoji that has a length of 2 (flags) WHEN we check if it's a single emoji THEN it should return true
```
Comment (tgolen) [inline tests/unit/EmojiTest.js:58]:
```suggestion
        // GIVEN an emoji that has a length of 2 (flags) WHEN we check if it's a single emoji THEN it should return true
```

Comment (tgolen) [reply tests/unit/EmojiTest.js:58]: A lot of the comments still reference "a single emoji". Can you please update the comments to speak instead about only containing emojis?

Also, can you please add tests that confirm:
- Multiple emojis, with no text (and any amount of whitespace), returns true
- single emojis, with some text, returns false
- multiple emojis, with some text, returns false

**Review by tgolen — CHANGES_REQUESTED** [review]

File: tests/unit/EmojiTest.js (hunk 1 above)
Comment (akshayasalvi) [reply tests/unit/EmojiTest.js]: @tgolen I've added tests and fixed your comments.

**Comment by akshayasalvi** [comment]
I accidentally removed @mananjadhav from the review, by request a second review from @tgolen

File: tests/unit/EmojiTest.js [hunk 2]
```diff
@@ -25,40 +25,52 @@ describe('EmojiTest', () => {
         expect(emojiMatched).toBe(true);
     });
 
-    it('matches single emojis variants for size', () => {
-        // GIVEN an emoji that has the default Unicode representation WHEN we check if it's a single emoji THEN it should return true
+    it('matches only emojis variants for size', () => {
```
Comment (mananjadhav) [inline tests/unit/EmojiTest.js:28]:
```suggestion
    it('matches emojis for different variants', () => {
```

I don't think the size is relevant anymore now.

File: src/libs/EmojiUtils.js [hunk 3]
```diff
@@ -254,11 +235,11 @@ function suggestEmojis(text, limit = 5) {
...
 
 export {
-    isSingleEmoji,
     getDynamicHeaderIndices,
     mergeEmojisWithFrequentlyUsedEmojis,
     addToFrequentlyUsedEmojis,
     containsOnlyEmojis,
     replaceEmojis,
     suggestEmojis,
+    trimEmojiUnicode,
```
Comment (mananjadhav) [inline src/libs/EmojiUtils.js:244]: This wasn't exported earlier and now we're exporting? Can't find its usage in the PR.

**Review by mananjadhav — CHANGES_REQUESTED** [review]
Minor comments, rest looks good to me.

File: tests/unit/EmojiTest.js (hunk 2 above)
Comment (akshayasalvi) [reply tests/unit/EmojiTest.js:28]: added

File: src/libs/EmojiUtils.js (hunk 3 above)
Comment (tgolen) [reply src/libs/EmojiUtils.js:244]: I was curious about this too

Comment (akshayasalvi) [reply src/libs/EmojiUtils.js:244]: This isn't used anymore in the code, but I used it in the script tagged in `emojis.js`. So kept it as a reference.

File: tests/unit/EmojiTest.js [hunk 4]
```diff
@@ -25,40 +25,52 @@ describe('EmojiTest', () => {
...
+        expect(EmojiUtils.containsOnlyEmojis('😄 is smiley')).toBe(false);
+
+        // GIVEN an input WHEN we check text and multiple emojis, THEN it should return false
+        expect(EmojiUtils.containsOnlyEmojis('Hi 😄👋')).toBe(false);
+
+        // GIVEN an input WHEN we only multiple emojis, THEN it should return true
+        expect(EmojiUtils.containsOnlyEmojis('😄👋')).toBe(true);
+
+        // GIVEN an input WHEN we check only multiple emojis with additional whitespace, THEN it should return false
+        expect(EmojiUtils.containsOnlyEmojis('😄  👋')).toBe(true);
```
[the remark this answers is not in the document]
Comment (tgolen) [reply tests/unit/EmojiTest.js:73]: Also, while you're at it, we had an internal discussion a while back and we agreed to stop doing GIVEN/WHEN/THEN in all caps. You can just capitalize them as they would be normally in a sentence

**Comment by akshayasalvi** [comment]
@mananjadhav PR updated

File: tests/unit/EmojiTest.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (akshayasalvi) [reply tests/unit/EmojiTest.js:78]: Okay I already updated, So I'll update all them in regular case?

[the remark this answers is not in the document]
Comment (akshayasalvi) [reply tests/unit/EmojiTest.js:78]: added 3 cases, and changed the cases for `GIVEN/WHEN/THEN`

**Review by mananjadhav — APPROVED** [review]

**Review by tgolen — APPROVED** [review]

**Comment by tgolen** [comment]
All checklists are complete, removing the emergency label

**Pull request merged**